source/CleoBase.cpp

```diff
@@ -111,42 +111,77 @@ namespace CLEO
-        CodeInjector.ReplaceFunction(OnScmInit3, VersionManager.TranslateMemoryAddress(MA_CALL_INIT_SCM3), &ScmInit3_Orig);
+            CodeInjector.ReplaceFunction(OnScmInit1, VersionManager.TranslateMemoryAddress(MA_CALL_INIT_SCM1), &ScmInit1_Orig);
+            CodeInjector.ReplaceFunction(OnScmInit2, VersionManager.TranslateMemoryAddress(MA_CALL_INIT_SCM2), &ScmInit2_Orig);
+            CodeInjector.ReplaceFunction(OnScmInit3, VersionManager.TranslateMemoryAddress(MA_CALL_INIT_SCM3), &ScmInit3_Orig);
 
-        CodeInjector.ReplaceFunction(OnGameShutdown, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_SHUTDOWN), &GameShutdown);
+            CodeInjector.ReplaceFunction(OnGameShutdown, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_SHUTDOWN), &GameShutdown);
 
-        CodeInjector.ReplaceFunction(OnGameRestart1, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_1), &GameRestart1);
-        CodeInjector.ReplaceFunction(OnGameRestart2, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_2), &GameRestart2);
-        CodeInjector.ReplaceFunction(OnGameRestart3, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_3), &GameRestart3);
+            CodeInjector.ReplaceFunction(OnGameRestart1, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_1), &GameRestart1);
+            CodeInjector.ReplaceFunction(OnGameRestart2, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_2), &GameRestart2);
+            CodeInjector.ReplaceFunction(OnGameRestart3, VersionManager.TranslateMemoryAddress(MA_CALL_GAME_RESTART_3), &GameRestart3);
+
+            OpcodeSystem.Init();
+            PluginSystem.LoadPlugins();
+        }
+        else // second phase of initialization
+        {
+            if (m_bLateStarted) return; // already started
+            m_bLateStarted = true;
+
+            // install "drawing finished" hook
+            // this is an empty function in original game, but some other mods have exactly same idea of hooking it
+            // wait for game started then install our hook
+            BYTE* address = VersionManager.TranslateMemoryAddress(MA_FLUSH_OBRS_PRINTFS);
+            switch (*address)
+            {
+                case OP_JMP: // there is hook!
+                    CodeInjector.ReplaceFunction(OnFlushObrsPrintfs, address, &FlushObrsPrintfs); // keep address of original hook
+                    break;
 
-        CodeInjector.ReplaceFunction(OnDrawingFinished, 0x00734640); // nullsub_63 - originally something like renderDebugStuff?
+                default:
+                    LOG_WARNING(0, "Unknown 'drawing finished' hook detected! Overwriting..."); // warning and install our instead
 
-        OpcodeSystem.Init();
-        PluginSystem.LoadPlugins();
+                case OP_RET: // still original code or some unknown hack
+                    CodeInjector.ReplaceFunction(OnFlushObrsPrintfs, address);
+                    break;
+            }
+        }
     }
```

Comment:
this should be a separate method, no need to increase complexity of an already long function

Reply:
Have single initialization function instead of two. After more thinking it probably be made with variable "int initializationStage", and incremented. So next stage can not be performed until previous was executed.

Reply:
no need to over-engineer simple stuff. please create a new method called `setupOnDrawingHook` and call it separately from the core init.

Reply:
Updated. There will be more things to initialize late. I had exact same need to start ModLoader system after all other asi plugins are loaded.

Hooking read string param function also can be considered to perform later.
Maybe that whole logic should be wrapped into SmartHook class.

Reply:
Updated